Makes search action cancelable by task management API

imotov opened the pull request:
Long running searches now can be cancelled using standard task cancellation mechanism.

nik9000 commented:
@imotov, want me to review it?

imotov commented:
@nik9000 sure - you from Task Management perspective and we will need somebody closer to Lucene ideally.

nik9000 commented on core/src/main/java/org/elasticsearch/action/search/SearchTask.java:
```diff
@@ -0,0 +1,34 @@
+    public SearchTask(long id, String type, String action, String description, TaskId parentTaskId) {
+        super(id, type, action, description, parentTaskId);
+    }
+
```
It'd be cool to be able to list the phase and/or which shards you are waiting for. You could put all kinds of cool stuff in here one day! But for now this seems like the right thing to do.

imotov replied:
Yes, that's the plan, but I didn't want to clump everything into a single PR.

nik9000 commented on core/src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -227,10 +229,11 @@ protected void doClose() {
 
-    public DfsSearchResult executeDfsPhase(ShardSearchRequest request) throws IOException {
+    public DfsSearchResult executeDfsPhase(ShardSearchRequest request, SearchTask task) throws IOException {
         final SearchContext context = createAndPutContext(request);
```
Maybe `createContext` should take the `SearchTask` as an argument? That'd make it very difficult to forget to set it.

jpountz replied:
+1

imotov replied:
See @nik9000's comment bellow, this doesn't seem to be a good way to handle it since the task can change over time.

nik9000 commented on core/src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -290,11 +294,12 @@ public QuerySearchResultProvider executeQueryPhase(ShardSearchRequest request) t
         SearchOperationListener operationListener = context.indexShard().getSearchOperationListener();
         context.incRef();
         try {
+            context.setTask(task);
```
Oooh! Yeah! That makes sense - scrolls contexts will outlast that task that spawns them and need to get a new task.

I haven't read the whole PR yet, but I wonder if we should clear task from the context between requests? If we don't then we maintain a reference to it after the task manager and that seems a bit weird. Changes to it wouldn't be useful to anyone.

nik9000 commented on core/src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -633,6 +643,7 @@ private void cleanContext(SearchContext context) {
         try {
             assert context == SearchContext.current();
             context.clearReleasables(Lifetime.PHASE);
+            context.setTask(null);
```
And here is the clear. Cool.

nik9000 commented on core/src/main/java/org/elasticsearch/search/dfs/DfsPhase.java:
```diff
@@ -58,6 +59,9 @@ public void execute(SearchContext context) {
             IndexReaderContext indexReaderContext = context.searcher().getTopReaderContext();
             for (int i = 0; i < terms.length; i++) {
+                if(context.isCancelled()) {
+                    throw new SearchContextException(context, "Cancelled");
```
This one is upper case and the one below is lower case.

nik9000 commented on core/src/main/java/org/elasticsearch/search/internal/SearchContext.java:
```diff
@@ -114,6 +116,18 @@ public ParseFieldMatcher parseFieldMatcher() {
         return parseFieldMatcher;
     }
 
+    public void setTask(SearchTask task) {
```
It'd be nice to have javadoc for these, especially given that task changes for scrolls.

nik9000 commented on core/src/main/java/org/elasticsearch/search/internal/SearchContext.java:
```diff
@@ -102,6 +103,7 @@ public static SearchContext current() {
     private Map<Lifetime, List<Releasable>> clearables = null;
     private final AtomicBoolean closed = new AtomicBoolean(false);
     private InnerHitsContext innerHitsContext;
+    private SearchTask task;
```
It is probably worth some javadoc explaining why this isn't `final` and why it is ok that it isn't `volatile`.

nik9000 commented on core/src/main/java/org/elasticsearch/search/query/CancellableCollector.java:
```diff
@@ -0,0 +1,62 @@
+
+        @Override
+        public void collect(int doc) throws IOException {
+            if (searchContext.isCancelled()) {
```
It is worth benchmarking this because it adds a volatile read. If it is too heavy we can make the check less frequent I imagine.

nik9000 commented on core/src/main/java/org/elasticsearch/search/query/QueryPhase.java:
```diff
@@ -361,6 +361,15 @@ public TopDocs call() throws Exception {
             }
 
+            if (collector != null && searchContext.getTask() != null) {
+                final Collector child = collector;
```
A null task here is pretty weird, right?

nik9000 commented:
@jpountz, would you like to have a look at this from a Lucene perspective?

jpountz commented on core/src/main/java/org/elasticsearch/transport/TransportService.java:
```diff
@@ -447,6 +448,27 @@ public void removeConnectionListener(TransportConnectionListener listener) {
+            handler.handleException(new TransportException(ex));
+            return;
+        }
+        sendRequest(node, action, request, options, handler);
```
nit: I'd prefer that we move the sendRequest call to the try block and remove the return from the catch block

jpountz commented on core/src/main/java/org/elasticsearch/search/internal/SearchContext.java:
```diff
@@ -114,6 +120,18 @@ public ParseFieldMatcher parseFieldMatcher() {
+    }
+
+    public boolean isCancelled() {
+        return task != null && task.isCancelled();
```
I don't think task should ever be null at that point? Should we just assert that `task` is not null?

jpountz commented on core/src/main/java/org/elasticsearch/search/query/QueryPhase.java:
```diff
@@ -361,6 +361,15 @@ public TopDocs call() throws Exception {
 
+            if (collector != null) {
+                final Collector child = collector;
+                collector = new CancellableCollector(searchContext, collector);
```
This part can cause significant slow downs. There may be ways that we can make it better, eg. by doing it at the bulk scorer level so that we can perform the check less often. Can we remove it for now and work on making this part cancellable in another PR. Or at least make it an opt-in for now with a setting?

imotov replied:
@jpountz  what do you think about this 662e1f5e3a2e9db89511749fe104e922edbb86ad?

jpountz replied:
I'm not a fan of the name (but I don't have better ideas) but this looks good to me.

jpountz commented on core/src/main/java/org/elasticsearch/search/query/CancellableCollector.java:
```diff
@@ -0,0 +1,68 @@
+    private final SearchContext searchContext;
+    private final boolean leafLevel;
+
+    public CancellableCollector(SearchContext searchContext, Collector in) {
```
Can we avoid taking a SearchContext as it makes this class hard to unit test, maybe we could eg. replace the search context with a `BooleanSupplier` and use a method reference? Then can you add some basic unit tests to this class?

jpountz commented on core/src/main/java/org/elasticsearch/search/profile/query/CollectorResult.java:
```diff
@@ -45,6 +45,7 @@
     public static final String REASON_SEARCH_MIN_SCORE = "search_min_score";
     public static final String REASON_SEARCH_MULTI = "search_multi";
     public static final String REASON_SEARCH_TIMEOUT = "search_timeout";
+    public static final String REASON_SEARCH_CANCELED = "search_canceled";
```
should we be consistent with the number of `l`? there is only one here while you put 2 in `cancellation`? (feel free to ignore, it could easily a special case in english that I don't know about!)

jpountz commented on core/src/main/java/org/elasticsearch/search/internal/SearchContext.java:
```diff
@@ -220,6 +238,11 @@ public InnerHitsContext innerHits() {
 
     public abstract void terminateAfter(int terminateAfter);
 
+    // Indicates if the current index should perform frequent low level search cancellation check
```
Can you make it an actual javadoc and explain the trade-off?

imotov commented:
@jpountz I have implemented the changes that you have requested and rebased it against the current master since it quite a few things has changed. Would you mind taking another look when you have a chance?

jpountz commented on core/src/main/java/org/elasticsearch/search/query/CancellableCollector.java:
```diff
@@ -0,0 +1,76 @@
+    private final boolean leafLevel;
+
+    /**
+     * Constractor
```
construction contractor? :)

jpountz commented on core/src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -546,6 +567,7 @@ final SearchContext createContext(ShardSearchRequest request, @Nullable Engine.S
                 keepAlive = request.scroll().keepAlive().millis();
             }
             context.keepAlive(keepAlive);
+            context.lowLevelCancellation(lowLevelCancellation);
```
why do we only do it in the fetch phase?

jpountz replied:
sorry I just realized this is the createContext method and not executeFetchPhase as I initially thought.

then I'm wondering if we could have the lowLevelCancellation only on DefaultSearchContext rather than the SearchContext base class?

imotov replied:
done

jpountz commented:
The search part looks good to me.

nik9000 commented on core/src/main/java/org/elasticsearch/transport/TaskAwareTransportRequestHandler.java:
```diff
@@ -0,0 +1,30 @@
+/**
+ * Transport request handlers that is using task context
+ */
+public abstract class TaskAwareTransportRequestHandler<T extends TransportRequest> implements TransportRequestHandler<T>  {
```
I guess you could make this a final class that takes a functional interface as its only parameter and runs it and you could still use lambdas for it. My instincts are that that is marginally better from a design standpoint (composition vs inheritance) and a little easier to read but I'm not sure.

imotov replied:
This sounds like an interesting idea, but I think we should do it on the TransportRequestHandler then and should, probably, be another PR.

nik9000 replied:
++

nik9000 commented on core/src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -107,6 +109,8 @@
         Setting.positiveTimeSetting("search.default_keep_alive", timeValueMinutes(5), Property.NodeScope);
     public static final Setting<TimeValue> KEEPALIVE_INTERVAL_SETTING =
         Setting.positiveTimeSetting("search.keep_alive_interval", timeValueMinutes(1), Property.NodeScope);
+    public static final Setting<Boolean> LOW_LEVEL_CANCELLATION_SETTING =
```
Can haz javadoc?

nik9000 commented on core/src/main/java/org/elasticsearch/search/query/CancellableCollector.java:
```diff
@@ -0,0 +1,76 @@
+
+    /**
+     * Constructor
+     * @param cancelled supplier of the cancellation flag
```
Maybe mention that this amounts to a volatile read.

imotov replied:
It's supplier implementation detail, isn't it? It seems to be wrong place to say something like this. However, the information about how often it will be called can be useful here. I will add that.

nik9000 replied:
Something like "This class assumes that the supplier is fast, with performance on the order of a volatile read." would give a lot of context to the decisions around how to use the Supplier.

nik9000 commented on core/src/test/java/org/elasticsearch/search/SearchCancellationIT.java:
```diff
@@ -0,0 +1,291 @@
+import static org.hamcrest.Matchers.greaterThan;
+
+/**
+ */
```
Can you remove the empty javdoc? We are going to fail the build on those at some point....

nik9000 commented on core/src/test/java/org/elasticsearch/search/SearchCancellationIT.java:
```diff
@@ -0,0 +1,291 @@
+
+/**
+ */
+@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.SUITE)
```
Can you explain why this is needed in a comment? I'm never sure when to use this and when not to so I figure it'd be nice to have it explained.

imotov replied:
Because I want nodes with specific settings and I don't want other tests to run with these settings.

nik9000 replied:
Then could you do something like:
```
@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.SUITE) // Changes settings
```

That'd make it clear why it is needed so folks don't go copying it into their tests without knowing why... 👼

nik9000 commented on core/src/test/java/org/elasticsearch/search/SearchCancellationIT.java:
```diff
@@ -0,0 +1,291 @@
+
+    private void indexTestData() {
+        for (int i = 0; i < 100; i++) {
+            client().prepareIndex("test", "type", Integer.toString(i)).setSource("field", "value").get();
```
Is this any quicker if you use bulks? I tend to do that out of habit.

nik9000 commented on core/src/test/java/org/elasticsearch/search/SearchCancellationIT.java:
```diff
@@ -0,0 +1,291 @@
+            client().prepareIndex("test", "type", Integer.toString(i)).setSource("field", "value").get();
+            if (i % 10 == 0) {
+                // Make sure we have a few segments
+                client().admin().indices().prepareRefresh("test").get();
```
Probably worth asserting that they return success. I've never seen them not do so but if they ever start then this test won't have many segments and will fail in strange ways to anyone trying to debug it.

nik9000 commented on core/src/test/java/org/elasticsearch/search/SearchCancellationIT.java:
```diff
@@ -0,0 +1,291 @@
+
+        logger.info("Cancelling search");
+        CancelTasksResponse cancelTasksResponse = client().admin().cluster().prepareCancelTasks().setTaskId(searchTask.getTaskId()).get();
+        assertThat(cancelTasksResponse.getTasks().size(), equalTo(1));
```
I find that `equalTo` is almost always a bad choice. In this case I think `assertThat(cancelTaskResponse.getTask(), hasSize(1));` will do the same thing but have much better error reporting. That way you get to see all the tasks when there are too many. Same for the above assertion.

nik9000 commented on core/src/test/java/org/elasticsearch/search/SearchCancellationIT.java:
```diff
@@ -0,0 +1,291 @@
+        try {
+            SearchResponse response = searchResponse.actionGet();
+            logger.info("Search response {}", response);
+            if (response.getFailedShards() == 0) {
```
`assertEquals("At least one shard should have failed", 0, response.getFailedShards());`?

nik9000 commented on core/src/test/java/org/elasticsearch/search/SearchCancellationIT.java:
```diff
@@ -0,0 +1,291 @@
+        ensureSearchWasCancelled(scrollResponse);
+
+        // Occasionally, cancellation might not hit all shards, leaving the scroll id hanging on these shards
+        client().prepareClearScroll().addScrollId(scrollId).get();
```
Does this happen because the scroll doesn't hit docs in that segment so it can't be blocked?

Should canceling a scroll request nuke the scroll id? If you tried to pump it again would it be busted? Maybe something for a followup though.

nik9000 commented on core/src/test/java/org/elasticsearch/search/SearchCancellationIT.java:
```diff
@@ -0,0 +1,291 @@
+
+    @Override
+    protected Settings nodeSettings(int nodeOrdinal) {
+        return Settings.builder().put(SearchService.LOW_LEVEL_CANCELLATION_SETTING.getKey(), true).build();
```
I wonder if it is worth only turning this on for the tests that needs it so it is obvious which ones do?

imotov replied:
Good catch! That was supposed to `randomBoolean()`. I replaced it with `true` while I was chasing a specific failure and forgot to flip it back on. I am unit testing that lowLevel is getting applied properly, so all tests should pass with either setting.

nik9000 replied:
Awesome.

nik9000 commented on core/src/test/java/org/elasticsearch/search/SearchCancellationTests.java:
```diff
@@ -0,0 +1,120 @@
+        leafCollector.collect(0);
+        cancelled.set(true);
+        try {
+            leafCollector.collect(1);
```
`expectThrows`?

nik9000 commented on core/src/test/java/org/elasticsearch/search/SearchCancellationTests.java:
```diff
@@ -0,0 +1,120 @@
+        for(int i=0; i<1000; i++) {
+            leafCollector.collect(0);
+        }
+        logger.info("took: {}", System.nanoTime() - start);
```
If we keep this maybe we should rename it so it doesn't run by default? And javadoc it? It is fairly obvious what it is for but I expect it isn't worth doing it during a regular test run because no one is reading it....

imotov replied:
Leftovers, deleted.

nik9000 commented on docs/reference/search.asciidoc:
```diff
@@ -92,6 +92,18 @@ timeout. The setting key is `search.default_search_timeout` and can be
+[[global-search-cancellation]]
+== Search Cancellation
+
+Long running searches can be cancelled using standard <<task-cancellation,task cancellation>>
```
s/Long running search/Searchs/ ?

nik9000 commented on docs/reference/search.asciidoc:
```diff
@@ -92,6 +92,18 @@ timeout. The setting key is `search.default_search_timeout` and can be
+mechanism. By default, a running search only checks if it is cancelled or
+not on segment boundaries, therefore the cancellation can be delayed by large
+segments. The search cancellation responsiveness can be improved by setting
+the dynamic cluster-level setting `search.low_level_cancellation` to `true`.
```
Mention that it only effects searches after the change to the setting.

nik9000 replied:
Are we thinking of automatically opting certain thing into this setting? Maybe we can have search benchmarks that compare?

nik9000 commented on docs/reference/search/profile.asciidoc:
```diff
@@ -150,7 +157,8 @@ This will yield the following result:
 // TESTRESPONSE[s/"next_doc": 2852/"next_doc": $body.profile.shards.0.searches.0.query.0.children.1.breakdown.next_doc/]
-// TESTRESPONSE[s/"time": "0.06989100000ms"/"time": $body.profile.shards.0.searches.0.collector.0.time/]
+// TESTRESPONSE[s/"time": "0.3043110000ms"/"time": $body.profile.shards.0.searches.0.collector.0.time/]
+// TESTRESPONSE[s/"time": "0.03227300000ms"/"time": $body.profile.shards.0.searches.0.collector.0.children.0.time/]
```
I'm sorry about this nightmare.....

nik9000 replied:
I'd love some way to say "just match anything in this position that looks like X" so you don't have to describe the whole path....

nik9000 commented on docs/reference/search/profile.asciidoc:
```diff
@@ -390,21 +398,30 @@ Looking at the previous example:
-We see a single collector named `SimpleTopScoreDocCollector`.  This is the default "scoring and sorting" Collector
-used by Elasticsearch.  The `"reason"` field attempts to give a plain english description of the class name.  The
+We see a single collector named `SimpleTopScoreDocCollector` wrapped into `CancellableCollector`. `SimpleTopScoreDocCollector` is the default "scoring and sorting"
+Collector used by Elasticsearch.  The `"reason"` field attempts to give a plain english description of the class name.  The
```
`Collector`?

nik9000 replied:
Like, with the backticks.
